pandas/core/groupby/groupby.py

```diff
@@ -4767,8 +4767,20 @@ def rank(
             result = self._python_apply_general(
                 f, self._selected_obj, is_transform=True
             )
+            if method != "average" and not pct:
+                temp = result
+                temp = temp.convert_dtypes()
+                return temp
             return result
-
+        if method != "average" and not pct:
+            temp = self._cython_transform(
+                "rank",
+                numeric_only=False,
+                axis=axis,
+                **kwargs,
+            )
+            temp = temp.convert_dtypes()
```

Comment:
This will convert e.g. float64 to Int64 I think, which we don't want. I also don't think you can do something like `result.astype("int64")` as you _will_ want e.g. Int64 input to remain Int64 output. I'd hazard a guess it'd be better for the change to be deeper than this, perhaps within `pandas._libs.groupby.group_rank` itself.

Reply:
Sorry for the confusion, but I thought that converting float64 into Int64 in the specified case of `method != "average" and not pct` was what was described in the Issue. Is this assumption incorrect?

Reply:
That is correct, but this isn't going to give you `int64`, it gives you `Int64`. Like you mentioned in the OP:

> E [left]: Int64
> E [right]: float64

`Int64` is the extension array: https://pandas.pydata.org/pandas-docs/dev/user_guide/integer_na.html

Reply:
From the OP in the linked issue:

> In these cases, it would be more natural to return int64 instead of float64